### DISCUSSION TITLE
Publishing events default strategy - interrupted when error occured inside handler.

### DISCUSSION THREAD

**Discussion by LukasKuchta** [LuckyPennySoftware/MediatR#947] [Q&A] [answered]
Hi, i'm trying to get right answer. 

What is my goal. I want to consume notification all handlers if it is possible and when not a retry sending. 

**Default foreach strategy**
When i have 5 notification handlers and for example 2 is broken (unhandled expcetion here) other handlers in sequence are not invoked becouse publising loop is break. It is expected behaviour for production? If there is any retry mechanism for sending notifaction more that once every attempt fail in second handler and rest is untouched. 

Ok. Is there a possible to create own strategy. Made loop interate over all items and save expcetion when happen. Throw it after loop is end. 

Or is there any better solution for this ?

**Comment by jbogard** [comment, marked as answer]
You'll want to create your own publish strategy for this, which is easy to do. Just create a class that implements `INotificationPublisher` and register it in `AddMediatR` with the `NotificationPublisher` property.

Is this expected behavior? Yes. Is it desired? That's entirely up to you.

That being said, some folks prefer to handle that logic *inside* the handler using something like Polly. But it's very easily extensible to do whatever you want.

**Reply by LukasKuchta** [reply]
Thanks. I did do as you said.

```
public class UnbreakableForeachAwaitPublisher : INotificationPublisher
    {
        public async Task Publish(IEnumerable<NotificationHandlerExecutor> handlerExecutors, INotification notification, CancellationToken cancellationToken)
        {
            IList<Exception>? exceptionList = null;

            foreach (var handler in handlerExecutors)
            {
                try
                {
                    await handler.HandlerCallback(notification, cancellationToken).ConfigureAwait(false);
                }
                catch (Exception ex)
                {
                    if (exceptionList is null)
                    {
                        exceptionList = new List<Exception>;
                    }

                    exceptionList.Add(ex);
                }
            }

            if (exceptionList is not null)
            {
                throw new AggregateException(
                            "Encountered errors while trying to publish notifications.",
                            exceptionList);
            }
        }
    }
```
